docs/developer/python-openflow.rst

```diff
@@ -0,0 +1,206 @@
+
+How to use the python-openflow library
+======================================
+
+importing
++++++++++
+To use the python-openflow module you need to import its classes directly
+from its submodule.
+
+how to do it
+~~~~~~~~~~~~
+To create a v0x04 Hello message for example, you need to import its class
+from the ``pyof.v0x04.symmetric.hello`` submodule using the statement:
+
+.. code:: python
+
+  >>> from pyof.v0x01.symmetric.hello import Hello
+  >>> hello_message = Hello()
+
+how NOT to do it
+~~~~~~~~~~~~~~~~
+Importing only ``pyof``, a version subpackage like ``pyof.v0x01``,
+or even a message category subpackage like ``pyof.v0x01.symmetric`` doesn't
+provide you anything.
+
+.. code:: python
+
+  >>> import pyof
+  >>> pyof.v0x01.symmetric.hello.Hello()
+
+  AttributeError: module 'pyof' has no attribute 'v0x01'
+
+
+.. code:: python
+
+  >>> old_dir = dir() + ['old_dir']
+
+  >>> from pyof import *
+  >>> from pyof.v0x01 import *
+  >>> from pyof.v0x01.symmetric import *
+
+  >>> [i for i in dir() if i not in old_dir]
+  []
+
+nothing is imported...
+
+pyof Objects
+++++++++++++
+pyof objects are instances of GenericStruct or GenericType or other classes
+derived from these.
+They provide the methods ``pack`` ``unpack`` and ``get_size``.
+
+initialization
+++++++++++++++
+``GenericType``s accept as argument an initial value.
+``GenericStruct``s derivatives usually accept as argument initial values
+for its class attributes.
+
+packing
++++++++
+To pack a pyof object, simply call its ``pack()`` method, which returns a
+``bytes`` object of the binary representation.
+
+unpacking
++++++++++
+To unpack a buffer as pyof object, simply initialize a new object and
+call its ``unpack()`` method, passing the buffer as the argument. The
+buffer will be unpacked in place setting the attributes of the object
+instance used to make the call.
+
+
+New definitions
+===============
+``python-openflow`` provides some base classes and basic type classes meant to
+be used in a new message/struct class definition.
+
+Difference between "structs" and "messages"
++++++++++++++++++++++++++++++++++++++++++++
+Messages are GenericStruct derivatives who carry a header attribute containing
+an OpenFlow header struct.
+
+Unlike ``GenericStruct``s their unpack method accepts a buffer argument with
+its packed body (without the associated header).
+
+How to code a new struct/message
+++++++++++++++++++++++++++++++++
+To implement a new struct, you need to:
+- define a new class which derives from ``GenericStruct``;
+- define class attributes in pack order (with optional initial values) whose
+  type are pyof objects themselves.
+
+Struct example:
+
+.. code:: python
+
+    >>> from pyof.foundation.base import GenericStruct
+    >>> from pyof.foundation.basic_types import UBInt8, UBInt16
+
+    >>> class MyNewStruct(GenericStruct):
+            my_first_attribute = UBInt8(255)
+            my_second_attribute = UBInt16(0)
+            my_third_attribute = UBInt8(255)
```

Comment:
Add also the `__init__` class passing the `__init__` arguments to show, on the example, that the instance can be created with the desired values on the object instantiation.